src/lib/tree/binary/test/BinaryMerkleTree.t.sol

```diff
@@ -101,6 +101,36 @@ contract BinaryMerkleProofTest is DSTest {
         assertTrue(isValid);
     }
 
+    function testVerifyLeafTwoOfEight() external {
+        bytes32 root = 0xc1ad6548cb4c7663110df219ec8b36ca63b01158956f4be31a38a88d0c7f7071;
+        bytes32[] memory sideNodes = new bytes32[](3);
+        sideNodes[0] = 0xb413f47d13ee2fe6c845b2ee141af81de858df4ec549a58b7970bb96645bc8d2;
+        sideNodes[1] = 0x78850a5ab36238b076dd99fd258c70d523168704247988a94caa8c9ccd056b8d;
+        sideNodes[2] = 0x4301a067262bbb18b4919742326f6f6d706099f9c0e8b0f2db7b88f204b2cf09;
+
+        uint256 key = 1;
+        uint256 numLeaves = 8;
+        BinaryMerkleProof memory proof = BinaryMerkleProof(sideNodes, key, numLeaves);
+        bytes memory data = hex"02";
+        bool isValid = BinaryMerkleTree.verify(root, proof, data);
+        assertTrue(isValid);
+    }
+
+    function testVerifyLeafThreeOfEight() external {
+        bytes32 root = 0xc1ad6548cb4c7663110df219ec8b36ca63b01158956f4be31a38a88d0c7f7071;
+        bytes32[] memory sideNodes = new bytes32[](3);
+        sideNodes[0] = 0x4f35212d12f9ad2036492c95f1fe79baf4ec7bd9bef3dffa7579f2293ff546a4;
+        sideNodes[1] = 0x6bcf0e2e93e0a18e22789aee965e6553f4fbe93f0acfc4a705d691c8311c4965;
+        sideNodes[2] = 0x4301a067262bbb18b4919742326f6f6d706099f9c0e8b0f2db7b88f204b2cf09;
+
+        uint256 key = 2;
+        uint256 numLeaves = 8;
+        BinaryMerkleProof memory proof = BinaryMerkleProof(sideNodes, key, numLeaves);
+        bytes memory data = hex"03";
+        bool isValid = BinaryMerkleTree.verify(root, proof, data);
+        assertTrue(isValid);
```

Comment:
can we also include a test where we expect the verification to fail?

Reply:
[a8855d0](https://github.com/celestiaorg/blobstream-contracts/pull/253/commits/a8855d08ae9353acafad23028e983b1b8d581025)

Reply:
It was a good idea to add invalid proofs :+1: Found a bug :D

Reply:
actually not a bug, but an edge case:

Assume you have a merkle tree of 5 leaves. When creating the proof, if you set the `number of leaves = [5, 8]`, the proof going to pass even if `{6, 7, 8}` are invalid values because with those numbers of leaves, the verifier will behave the same. 

Verified with the implementation of tendermint and it also has the same behaviour.

Reply:
Could you expand on this? What does `number of leaves = [5, 8]` mean? That's a range, not a number.

Reply:
My bad, i meant any number from that range will result in a valid proof. The verifier won't complain about it.

So if you have a tree of 5 leaves, and in the proof you set the total number to 7, the proof is still considered valid

Reply:
https://github.com/celestiaorg/blobstream-contracts/issues/259